### PR TITLE
Add support for secrets in Go SDK

### PR DESCRIPTION
**Pull request by EvanBoyle**
Adds secret state to `pulumi.Output`, support for `additionalSecretOutputs`, and `config.Get/Try/RequireSecret*()`.

I uncovered some test bugs along the way in RPC tests both in pkg/resource and in the go sdk, and fixed them as a part of this change. 

Fixes https://github.com/pulumi/pulumi/issues/2820

### CODE REVIEW

File: sdk/go/pulumi/types.go [hunk 1]
```diff
@@ -389,11 +391,33 @@ func (o *OutputState) ApplyTWithContext(ctx context.Context, applier interface{}
...
 }
 
+// IsSecret returns a bool representing the secretness of the Output
+func (o *OutputState) IsSecret() bool {
+	return o.getState().secret
+}
+
+// SecretT wraps the input in an Output marked as secret. The result remains an output
+// property, and accumulates all implicated dependencies, so that resources can be properly tracked using a DAG.
+// This function does not block awaiting the value; instead, it spawns a Goroutine that will await its availability.
```
Comment (lukehoban) [inline sdk/go/pulumi/types.go]: I don't quite follow the comment above.  What does this function have to do with spawning goroutines?  What does it have to do with a DAG?

File: sdk/go/pulumi/types.go [hunk 2]
```diff
@@ -389,11 +391,33 @@ func (o *OutputState) ApplyTWithContext(ctx context.Context, applier interface{}
...
+func SecretT(input interface{}) Output {
```
Comment (lukehoban) [inline sdk/go/pulumi/types.go]: This function is morally very similar to `ToOutput`.  Is there a reason this is not `ToSecret` for consistency?

File: sdk/go/pulumi/types_test.go [hunk 3]
```diff
@@ -301,3 +309,86 @@ func TestToOutputInputAny(t *testing.T) {
...
+
+// Test that SecretT sets appropriate internal state and that IsSecret appropriately reads it.
+func TestSecrets(t *testing.T) {
+	s := SecretT(String("foo"))
+	// assert that secret is immediately secret
+	assert.True(t, s.IsSecret())
+
+	errChan := make(chan error)
+	resultChan := make(chan string)
+	secretChan := make(chan bool)
```
Comment (pgavlin) [inline sdk/go/pulumi/types_test.go]: You should be able to use `await` in this test.

File: sdk/go/pulumi/types.go [hunk 4]
```diff
@@ -654,6 +682,10 @@ func ToOutput(v interface{}) Output {
 // ToOutputWithContext returns an Output that will resolve when all Outputs contained in the given value have
 // resolved.
 func ToOutputWithContext(ctx context.Context, v interface{}) Output {
+	return toMaybeSecretOutputWithContext(ctx, v, false)
+}
+
+func toMaybeSecretOutputWithContext(ctx context.Context, v interface{}, forceSecret bool) Output {
```
Comment (pgavlin) [inline sdk/go/pulumi/types.go]: Nit: this is unnecessarily wordy. Let's let the parameter name speak for itself.

```suggestion
func toOutputWithContext(ctx context.Context, v interface{}, forceSecret bool) Output {
```

File: sdk/go/pulumi/types_test.go (hunk 3 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/types_test.go]: True, but going to leave these as-is.

File: sdk/go/pulumi/types.go (hunk 1 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/types.go]: These comments are copied from Apply. You're right, they don't make much sense here.

File: sdk/go/pulumi/types.go (hunk 2 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/types.go]: Good point. I've updated this.

Comment (EvanBoyle) [reply sdk/go/pulumi/types.go]: I'll update the programming model docs as well.

File: sdk/go/pulumi/context.go [hunk 5]
```diff
@@ -647,30 +656,38 @@ func (state *resourceState) resolve(dryrun bool, err error, inputs *resourceInpu
...
+		secret := false
+		if v.IsSecret() {
+			secret = true
+		}
+
 		// Allocate storage for the unmarshalled output.
 		dest := reflect.New(output.ElementType()).Elem()
-		if err = unmarshalOutput(v, dest); err != nil {
+		usecret, err := unmarshalOutput(v, dest)
+		secret = secret || usecret
```
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply sdk/go/pulumi/context.go]: You are correct, this whole block is redundant (removed):

```go
		secret := false
		if v.IsSecret() {
			secret = true
		}
```

File: sdk/go/pulumi/rpc.go [hunk 6]
```diff
@@ -405,90 +434,107 @@ func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) error {
...
+			return true, nil
+		}
+
+		secretValType := reflect.TypeOf(secretVal)
+		if !secretValType.AssignableTo(dest.Type()) {
+			return true, errors.Errorf("expected a %s, got an %v", dest.Type(), secretValType)
+		}
+
+		dest.Set(reflect.ValueOf(secretVal))
+		return true, nil
```
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply sdk/go/pulumi/rpc.go]: Good catch, definitely a bug.

File: sdk/go/pulumi/types.go (hunk 2 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/types.go]: https://github.com/pulumi/docs/pull/2533

File: sdk/go/pulumi/config/config_test.go [hunk 7]
```diff
@@ -149,3 +151,180 @@ func TestConfig(t *testing.T) {
...
+	for i := 0; i < 3; i++ {
+		select {
+		case err = <-errChan:
+			assert.Nil(t, err)
+			break
+		case r := <-result:
+			assert.Equal(t, "a string value", r)
+			break
+		}
+	}
```
Comment (lukehoban) [inline sdk/go/pulumi/config/config_test.go]: Is this style common in our Go tests?  Does it not work to just do the asserts in the `Apply`?

Comment (pgavlin) [reply sdk/go/pulumi/config/config_test.go]: I think in the past I've used a `WaitGroup` to accomplish this and either run the asserts in the `Apply` or leaked the resolved values to the outside and validated them there.

File: sdk/go/pulumi/rpc.go [hunk 8]
```diff
@@ -152,25 +169,26 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
+					return resource.PropertyValue{}, nil, false, errors.Errorf(cannotAwaitFmt, output)
 				}
 
 				// Await the output.
-				ov, known, err := output.await(context.TODO())
+				ov, known, outputSecret, err := output.await(context.TODO())
+				secret = outputSecret
 				if err != nil {
-					return resource.PropertyValue{}, nil, err
+					return resource.PropertyValue{}, nil, secret, err
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
					return resource.PropertyValue{}, nil, false, err
```

File: sdk/go/pulumi/rpc.go [hunk 9]
```diff
@@ -152,25 +169,26 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
 			if output, ok := input.(Output); ok {
 				if !await {
-					return resource.PropertyValue{}, nil, errors.Errorf(cannotAwaitFmt, output)
+					return resource.PropertyValue{}, nil, false, errors.Errorf(cannotAwaitFmt, output)
 				}
 
 				// Await the output.
-				ov, known, err := output.await(context.TODO())
+				ov, known, outputSecret, err := output.await(context.TODO())
+				secret = outputSecret
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]: Might as well move this below the `if err != nil` block.

File: sdk/go/pulumi/rpc.go [hunk 10]
```diff
@@ -206,16 +224,16 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
-			}), deps, nil
+			}), deps, secret, nil
 		case CustomResource:
 			deps = append(deps, v)
 
 			// Resources aren't serializable; instead, serialize a reference to ID, tracking as a dependency.
 			e, d, err := marshalInput(v.ID(), idType, await)
 			if err != nil {
-				return resource.PropertyValue{}, nil, err
+				return resource.PropertyValue{}, nil, secret, err
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
				return resource.PropertyValue{}, nil, false, err
```

File: sdk/go/pulumi/rpc.go [hunk 11]
```diff
@@ -261,22 +279,22 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
 				elem := rv.Index(i)
 				e, d, err := marshalInput(elem.Interface(), destElem, await)
 				if err != nil {
-					return resource.PropertyValue{}, nil, err
+					return resource.PropertyValue{}, nil, secret, err
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
					return resource.PropertyValue{}, nil, false, err
```

File: sdk/go/pulumi/rpc.go [hunk 12]
```diff
@@ -261,22 +279,22 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
 					arr = append(arr, e)
 				}
 				deps = append(deps, d...)
 			}
-			return resource.NewArrayProperty(arr), deps, nil
+			return resource.NewArrayProperty(arr), deps, secret, nil
 		case reflect.Map:
 			if rv.Type().Key().Kind() != reflect.String {
-				return resource.PropertyValue{}, nil,
+				return resource.PropertyValue{}, nil, secret,
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
				return resource.PropertyValue{}, nil, false,
```

File: sdk/go/pulumi/rpc.go [hunk 13]
```diff
@@ -287,14 +305,14 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
 				value := rv.MapIndex(key)
 				mv, d, err := marshalInput(value.Interface(), destElem, await)
 				if err != nil {
-					return resource.PropertyValue{}, nil, err
+					return resource.PropertyValue{}, nil, secret, err
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
					return resource.PropertyValue{}, nil, false, err
```

File: sdk/go/pulumi/rpc.go [hunk 14]
```diff
@@ -308,90 +326,101 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
 
 				fv, d, err := marshalInput(rv.Field(i).Interface(), destField.Type, await)
 				if err != nil {
-					return resource.PropertyValue{}, nil, err
+					return resource.PropertyValue{}, nil, secret, err
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
					return resource.PropertyValue{}, nil, false, err
```

File: sdk/go/pulumi/rpc.go [hunk 15]
```diff
@@ -308,90 +326,101 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
 				if !fv.IsNull() {
 					obj[resource.PropertyKey(tag)] = fv
 				}
 				deps = append(deps, d...)
 			}
-			return resource.NewObjectProperty(obj), deps, nil
+			return resource.NewObjectProperty(obj), deps, secret, nil
 		}
-		return resource.PropertyValue{}, nil, errors.Errorf("unrecognized input property type: %v (%T)", v, v)
+		return resource.PropertyValue{}, nil, secret, errors.Errorf("unrecognized input property type: %v (%T)", v, v)
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]:
```suggestion
		return resource.PropertyValue{}, nil, false, errors.Errorf("unrecognized input property type: %v (%T)", v, v)
```

File: sdk/go/pulumi/rpc.go [hunk 16]
```diff
@@ -308,90 +326,101 @@ func marshalInput(v interface{}, destType reflect.Type, await bool) (resource.Pr
...
 // unmarshalOutput unmarshals a single output variable into its runtime representation.
-func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) error {
+// returning a bool that indicates secretness
+func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) (bool, error) {
 	contract.Assert(dest.CanSet())
 
 	// Check for nils and unknowns. The destination will be left with the zero value.
 	if v.IsNull() || v.IsComputed() || v.IsOutput() {
-		return nil
+		return v.IsSecret(), nil
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]: This will always be `false`.

```suggestion
		return false, nil
```

File: sdk/go/pulumi/rpc.go [hunk 17]
```diff
@@ -405,90 +434,99 @@ func unmarshalOutput(v resource.PropertyValue, dest reflect.Value) error {
...
 		}
 
 		result := reflect.MakeMap(dest.Type())
+		secret := false
 		for k, e := range v.ObjectValue() {
 			elem := reflect.New(elemType).Elem()
-			if err := unmarshalOutput(e, elem); err != nil {
-				return err
+			esecret, err := unmarshalOutput(e, elem)
+			secret = secret || esecret
```
Comment (pgavlin) [inline sdk/go/pulumi/rpc.go]: Nit: move below the `if err != nil` block

File: sdk/go/pulumi/types.go [hunk 18]
```diff
@@ -389,11 +391,31 @@ func (o *OutputState) ApplyTWithContext(ctx context.Context, applier interface{}
 		}
 
 		// Fulfill the result.
-		result.fulfillValue(results[0], true, nil)
+		result.fulfillValue(results[0], true, secret, nil)
 	}()
 	return result
 }
 
+// IsSecret returns a bool representing the secretness of the Output
```
Comment (pgavlin) [inline sdk/go/pulumi/types.go]: Do we expose this in any other SDKs?

Comment (EvanBoyle) [reply sdk/go/pulumi/types.go]: Looks like I missed the `internal` annotation in the node sdk. Will make private. 

https://github.com/pulumi/pulumi/blob/7fff99d346f140d32a196aeaf1aab192d0e0791d/sdk/nodejs/output.ts#L41

**Pull request merged**